changed pls_.py to properly scale data, modified documents

OlaPaw opened the pull request:
1.  In many instances of the variable `scale` setting was ignored.  Data was always centered, and always scaled.  The data is now scaled appropriately when `scale=True` is set, and not scaled when it's `False`
2. The return from the `predict()` function was always de-centered, but never de-scaled, so that we got:
   `Ypred = Ypred + self.y_mean_`  for all cases
   instead of:
   `Ypred = (Ypred * self.y_std_) + self.y_mean_` when `scale=True`
3. In the nipals algorithm, either C and Q matrices are used for deflation modes (regression or canonical).  This is now implemented.  The #FIXME note is now not necessary, as `coef_` is calculated differently depending on those two cases.  The calculated `coef_` was also scaled unnecesarily by standard deviations of X and Y: this is now commented out.  Only matrices W and C are scaled by their norms in the inner loop.  The outer loop scales P, Q, and C (canonical and regression) matrices by `T.T*T` or `U.T*U`
4. Some documentation was changed to reflect code modifications.  Also, instructions to use `sklearn.preprocessing.scale` are included, as that could be a useful step for most users.
5. It should be noted that the example included in the comment sections now gives better predictions of Y.
6. if X and Y are scaled using `sklearn.preprocessing.scale` (or column mean centered and column variance scaled), the pls_ functions all return the SAME results when `scale=True` and  `scale=False` this was not the case in the previous version of the code

Commit:
changed pls_.py to properly scale data, modified documents
and added changes to canonical regression

ogrisel commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -367,8 +375,12 @@ def fit(self, X, Y):
-            self.coef_ = (1. / self.x_std_.reshape((p, 1)) * self.coef_ *
-                          self.y_std_)
+            # self.coef_ = (1. / self.x_std_.reshape((p, 1)) * self.coef_ *
+            #               self.y_std_)
```
I am not familiar with PLS but if this is no longer needed please delete those commented out lines.

ogrisel commented:
Some tests are failing. They might need to be updated or relaxed a bit I am not sure.

@arthurmensch can probably have a look once the ICML deadline is over (tomorrow).

Also please add a one or more new test to highlight the cases that have been fixed in this PR and that were not properly tested previously.

arthurmensch commented:
@OlaPaw are you still willing to work on this. Sorry for the review delay btw. as @ogrisel said, we'll need non regression test for the scaling bug in #6002. It would be great to see if it fixes #6279. Do you happen to work under windows ?

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -41,7 +42,8 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
     eps = np.finfo(X.dtype).eps
-    # Inner loop of the Wold algo.
+    # Inner loop of the Wold algo. Very similar to chemometrics::pls2_nipals
+    # in R.  Tolarence looks at changes in the X matrix rather than Y
```
Tolerance

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -64,15 +66,14 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
+        y_score = np.dot(Y, y_weights)
         x_weights_diff = x_weights - x_weights_old
-        if np.dot(x_weights_diff.T, x_weights_diff) < tol or Y.shape[1] == 1:
+        if np.sqrt(np.dot(x_weights_diff.T, x_weights_diff)) < tol or Y.shape[1] == 1:
```
How is that necessary ? Alignment with similar R package ?

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -104,7 +113,7 @@ def _center_scale_xy(X, Y, scale=True):
     # scale
     if scale:
-        x_std = X.std(axis=0, ddof=1)
+        x_std = X.std(axis=0, ddof=1)  # calculate variance (n-1)
```
Comment not required IMHO

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -428,11 +443,16 @@ def predict(self, X, copy=True):
         X -= self.x_mean_
-        X /= self.x_std_
+        # Scale
+        if self.scale is True:
```
if self.scale

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -191,7 +200,7 @@ class _PLS(six.with_metaclass(ABCMeta), BaseEstimator, TransformerMixin,
 
     coef_: array, [p, q]
-        The coefficients of the linear model: ``Y = X coef_ + Err``
+        The coefficients of the linear model: ``Y = X coef_ + Err``, often B
```
often B ? I don't understand

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -93,6 +94,14 @@ def _svd_cross_product(X, Y):
 def _center_scale_xy(X, Y, scale=True):
     """ Center X, Y and scale if the scale parameter==True
```
Center X, Y, and optionally scales it

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -93,6 +94,14 @@ def _svd_cross_product(X, Y):
+    and 1.
+
+    For more general case, use sklearn.preprocessing.scale()
+
```
I am not sure this long commont is needed

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -52,7 +54,7 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
         else:  # mode A
             # Mode A regress each X column on y_score
-            x_weights = np.dot(X.T, y_score) / np.dot(y_score.T, y_score)
+            x_weights = np.dot(X.T, y_score)
```
If I understand weel, this line caused y to be always scaled not matter of the parameter, causing
#6002, right ?

arthurmensch commented:
Quoting @rlgc79:

> 

Hi OlaPaw,

I've seen that in your pull request you have added many statements of the type:

if self.scale is True:
    Xr /= self.x_std_

I am not sure whether the "if" statement is necessary. If you look at the code that scales the data, we can see that x_std and y_std are arrays of ones if we decided for not scaling the data:

    if scale:
        x_std = X.std(axis=0, ddof=1)
        x_std[x_std == 0.0] = 1.0
        X /= x_std
        y_std = Y.std(axis=0, ddof=1)
        y_std[y_std == 0.0] = 1.0
        Y /= y_std
    else:
        x_std = np.ones(X.shape[1])
        y_std = np.ones(Y.shape[1])
    return X, Y, x_mean, y_mean, x_std, y_std

As a result, I believe that we can safely use Xr /= self.x_std_ in your pull request, without the if statement, whether we decide to
scale the data or not.


I think the changes you made to the ```.fit```, ```.predict``` ```_scale_x_y``` are cosmetic one. The real bug #6002 fixing is in ```_nipals_twoblocks_inner_loop```.

 I think your cosmetic change make the code more explicit but I'd like to have another option on this.

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -64,15 +66,14 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
         # 2.3 Update y_score: the Y latent scores
-        y_score = np.dot(Y, y_weights) / (np.dot(y_weights.T, y_weights) + eps)
-        # y_score = np.dot(Y, y_weights) / np.dot(y_score.T, y_score) ## BUG
+        y_score = np.dot(Y, y_weights)
```
Effective bug fix

OlaPaw replied:
Not quite, but the dot product becomes unity, so is not necessary

arthurmensch commented on sklearn/cross_decomposition/pls_.py:
```diff
@@ -64,15 +66,14 @@ def _nipals_twoblocks_inner_loop(X, Y, mode="A", max_iter=500, tol=1e-06,
         else:
             # Mode A regress each Y column on x_score
-            y_weights = np.dot(Y.T, x_score) / np.dot(x_score.T, x_score)
+            y_weights = np.dot(Y.T, x_score)
```
Effective bug fix

ogrisel commented:
@arthurmensch is @OlaPaw does not reply please feel free to go ahead and extract the bugfix in a separate new PR and please also enable the tests on appveyor as done in #6274 to check whether it fixes #6279.

OlaPaw commented:
Hi @arthurmensch  and @ogrisel,

Yes, I think you have the entire gist of the bug here.  If you don't mind, I'll take a look at this over the weekend and test with all my separate test cases.  I'm really new to github and debugging code, so thank you for your patience.  

I just found a lot of the PLS code somewhat cryptic as it doesn't seem to follow the typical conventions I've seen in literature.  

Finally, scikit-learn already has pre-scaling function and I'm wondering whether we should even scale / center anything in this function.

OlaPaw committed:
Update pls_.py

no need for separate check if scale is True, as self_std is set to ones if set to False
Fixed re-scaling and re-centering issues reported previously

OlaPaw commented:
Ok, I've removed the cosmetic changes, and just fixed the issues with improper scaling and centering.  I tried to leave the rest of the code as pure as possible.

ogrisel commented:
Can you please add a non-regression test and update the existing tests (there are some failures reported by travis).

lesteve commented:
> It would be great to see if it also fixes #6279

Unfortunately it does not, I just checked.

amueller commented:
@OlaPaw are you still working on this? It would be good to have.

jabenitez commented:
I'm now using PLSRegression. Pls let me know if I can help complete this PR. I got here after noticing my outputs of predict were out of scale.

amueller commented:
@jabenitez feel free to help with this by adding the appropriate tests. You can create a new PR based on this one.

lesteve commented:
@Gscorreia89 as a non PLS expert, it is hard to figure out what remains to be done on this one now that #7819 has been merged. Would you be kind enough to elaborate a bit?
